flybywiresim/a32nx
# feat(atsu): request pages

## Summary of Changes

This PR introduces [FANS C](https://i.stack.imgur.com/oVc9a.png) and integrates it into the ATSU.
The communication types are not simulated by IVAO, VATSIM or Hoppie, but the A32NX uses an internal lookup table to decide based on the current CPDLC station which mode is correct. FANS A is used as the fallback, because it is used in most areas.

Additional to this are the following ATC request pages on the MCDU implemented:
- FANS A
  - LAT REQUEST
  - VERT REQUEST
  - TEXT
- FANS B
  - REQUEST

## Screenshots (if necessary)



## References


### FANS menus
[1] 
![grafik](https://user-images.githubusercontent.com/64070348/153953640-ef66086f-ec1d-4aab-870d-dde6809ca13c.png)

[2]
![grafik](https://user-images.githubusercontent.com/64070348/153953670-15fe1980-2b6b-4234-b0c7-4a6c8cc50c14.png)

### Area overview for FANS A or FANS B:
![grafik](https://user-images.githubusercontent.com/64070348/153945628-93a8b676-bcee-464b-b73a-8b614850b889.png)

### FANS A: valid input values and ragnes
![grafik](https://user-images.githubusercontent.com/64070348/155705912-943e0fbb-0497-4937-87d3-7527709cf215.png)
![grafik](https://user-images.githubusercontent.com/64070348/155705942-0691bf5b-ab20-402c-88f5-b7eec6a01fcf.png)
![grafik](https://user-images.githubusercontent.com/64070348/155705966-31ace32a-1e0c-49ce-b0bf-56158535089c.png)
![grafik](https://user-images.githubusercontent.com/64070348/155706000-c1233407-3844-4dd3-a5ec-681e1f457009.png)
![grafik](https://user-images.githubusercontent.com/64070348/155706022-9e0eb5a3-e632-4ec8-b80f-480a64274822.png)
![grafik](https://user-images.githubusercontent.com/64070348/155706039-c5a5bd04-fe79-4d5a-b383-d845aaa0348a.png)
![grafik](https://user-images.githubusercontent.com/64070348/155706070-923e9d27-33f2-467a-835c-bf46efdad46d.png)

### FANS B: valid input values and ranges
![grafik](https://user-images.githubusercontent.com/64070348/155704518-caac654b-dd3c-4287-97bb-b9c11e87d348.png)
![grafik](https://user-images.githubusercontent.com/64070348/155704479-6a2fe931-eb15-4d03-b1b0-09557c8cea74.png)
![grafik](https://user-images.githubusercontent.com/64070348/155704548-1cedc2ec-815c-480d-adb7-945036df1194.png)

### FANS A + B: message elements
![grafik](https://user-images.githubusercontent.com/64070348/156252179-b465ac10-3531-4367-a6af-23f6627cfd10.png)
![grafik](https://user-images.githubusercontent.com/64070348/156252266-fd80c9cb-f7d5-4388-90c1-bed4a66d390f.png)
![grafik](https://user-images.githubusercontent.com/64070348/156252565-52aaab54-590c-435a-93fe-5b508fe93912.png)
![grafik](https://user-images.githubusercontent.com/64070348/156252656-1f2e5232-8bde-45ca-b751-fc27542124ac.png)

#### Extensions for requests
![grafik](https://user-images.githubusercontent.com/64070348/156253031-d331771a-cb8d-4bb5-9f6d-81e48abade4a.png)

### FANS A: message elements
![grafik](https://user-images.githubusercontent.com/64070348/156252212-f178edae-a480-43fc-a0bf-1ec3064cba09.png)
![grafik](https://user-images.githubusercontent.com/64070348/156252428-4d8524eb-6c69-4b12-8651-b9cbb60ebf60.png)
![grafik](https://user-images.githubusercontent.com/64070348/156252456-feb569b3-db50-455b-8d3a-427a6048dde3.png)
![grafik](https://user-images.githubusercontent.com/64070348/156252532-9fd6dc5f-bf81-49ee-a035-e59f054e403c.png)
![grafik](https://user-images.githubusercontent.com/64070348/156252617-91c28028-3d76-4a4b-9354-a1c1b83d2c0e.png)
![grafik](https://user-images.githubusercontent.com/64070348/156252683-313d1fc2-004c-4f06-8db5-00a0c63513ad.png)
![grafik](https://user-images.githubusercontent.com/64070348/156252739-49bd31bb-d302-4cff-9a0a-1397655b4211.png)
![grafik](https://user-images.githubusercontent.com/64070348/156252800-986ef46b-2aa3-41c3-80a7-585c310019ef.png)
![grafik](https://user-images.githubusercontent.com/64070348/156252883-14739fad-48b2-4ea6-b463-51111c5ab5e5.png)
![grafik](https://user-images.githubusercontent.com/64070348/156252921-1bee02ac-89d4-402d-85fb-506ac1e59d9e.png)

#### Extensions for requests
![grafik](https://user-images.githubusercontent.com/64070348/156253087-5a707822-bfd6-4951-9090-7bc72a9d3abc.png)
- DUE TO MEDICAL, DUE TO TECHNICAL are defined as urgent freetext elements

### FANS B: message elements




## Additional context



Discord username (if different from GitHub): Sven [de en]

## Testing instructions

- test ATC Menu layout for FANS A [1]
  - Check the different valid input ranges 
- test ATC Menu layout for FANS B [2]
  - Check the different valid input ranges- 
  - temporary ATC station that simulates FANS B: TESB (will be removed before the merge)
  - test the input ranges for the entries defined in the upper references section
- Send requests to the ATC station and check if they are created correct (reference is the Datalink Gold edition)
  - Message definition given in the references
- Check that multiple requests are shown as one message in the DCDU, but are sent as seperate messages to the ATC
- test that FANS A menu is changed to FANS B and vice-versa if a handoff is performed from one FANS station to an other (i.e. TESA to TESB)


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
